storage/store/chunks_queue.go

```diff
@@ -0,0 +1,175 @@
+package store
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/model/chunks"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/metrics"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/operation"
+)
+
+// ChunksQueue stores a queue of chunk locators that assigned to me to verify.
+// Job consumers can read the locators as job from the queue by index.
+// Chunk locators stored in this queue are unique.
+type ChunksQueue struct {
+	db                storage.DB
+	chunkLocatorCache *Cache[uint64, *chunks.Locator] // cache for chunk locators, indexed by job index
+	storing           *sync.Mutex
+}
+
+const JobQueueChunksQueue = "JobQueueChunksQueue"
+
+func newChunkLocatorCache() *Cache[uint64, *chunks.Locator] {
+	store := func(rw storage.ReaderBatchWriter, index uint64, locator *chunks.Locator) error {
+		// make sure the chunk locator is unique
+		err := operation.InsertChunkLocator(rw.Writer(), locator)
+		if err != nil {
+			return fmt.Errorf("failed to insert chunk locator: %w", err)
+		}
+
+		err = operation.InsertJobAtIndex(rw.Writer(), JobQueueChunksQueue, index, locator.ID())
+		if err != nil {
+			return fmt.Errorf("failed to set job index for chunk locator queue at index %v: %w", index, err)
+		}
+
+		return nil
+	}
+
+	retrieve := func(r storage.Reader, index uint64) (*chunks.Locator, error) {
+		var locatorID flow.Identifier
+		err := operation.RetrieveJobAtIndex(r, JobQueueChunksQueue, index, &locatorID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve chunk locator in queue: %w", err)
+		}
+
+		var locator chunks.Locator
+		err = operation.RetrieveChunkLocator(r, locatorID, &locator)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve locator for chunk id %v: %w", locatorID, err)
+		}
+
+		return &locator, nil
+	}
+	return newCache(metrics.NewNoopCollector(), "",
+		withLimit[uint64, *chunks.Locator](DefaultCacheSize),
+		withStore(store),
+		withRetrieve(retrieve))
+}
+
+// NewChunkQueue will initialize the underlying badger database of chunk locator queue.
+func NewChunkQueue(db storage.DB) *ChunksQueue {
+	return &ChunksQueue{
+		db:                db,
+		chunkLocatorCache: newChunkLocatorCache(),
+		storing:           &sync.Mutex{},
+	}
+}
+
+// Init initializes chunk queue's latest index with the given default index.
+// It returns (false, nil) if the chunk queue is already initialized.
+// It returns (true, nil) if the chunk queue is successfully initialized.
+func (q *ChunksQueue) Init(defaultIndex uint64) (bool, error) {
+	q.storing.Lock()
+	defer q.storing.Unlock()
+
+	_, err := q.LatestIndex()
+	// the latest index should not exist,
+	// if the latest index is not found, initialize it with the default index
+	if errors.Is(err, storage.ErrNotFound) {
```

Comment:
you can reduce nesting by reversing the cases.

`if err== nil {return}`
`if !error.Is()`{return err}
`otherwise...`